pkg/controller/nodecsrapprover/node_csr_approver.go

```diff
@@ -0,0 +1,169 @@
+			return nil
+		}
+		return err
+	}
+	klog.V(4).Infof("Reconciling CSR %s", csr.ObjectMeta.Name)
+
+	// If CSR is approved, skip it
+	for _, condition := range csr.Status.Conditions {
+		if condition.Type == certificatesv1beta1.CertificateApproved {
+			klog.V(4).Infof("CSR %s already approved, skipping reconciling", csr.ObjectMeta.Name)
+			return nil
+		}
+	}
+
+	// Ensure system:nodes is in groups
+	if !sets.NewString(csr.Spec.Groups...).Has("system:nodes") {
+		klog.V(4).Infof("Skipping reconciling CSR '%s' because 'system:nodes' is not in its groups", csr.ObjectMeta.Name)
+		return nil
+	}
+
+	// Check are present usages matching allowed usages
+	if len(csr.Spec.Usages) != 3 {
+		klog.V(4).Infof("Skipping reconciling CSR '%s' because it has not exactly three usages defined", csr.ObjectMeta.Name)
+		return nil
+	}
+	for _, usage := range csr.Spec.Usages {
+		if !isUsageInUsageList(usage, allowedUsages) {
+			klog.V(4).Infof("Skipping reconciling CSR '%s' because its usage (%v) is not in the list of allowed usages (%v)",
+				csr.ObjectMeta.Name, usage, allowedUsages)
+			return nil
+		}
+	}
+	// Validate the CSR request
+	if err := validateCSRRequest(csr); err != nil {
+		klog.V(4).Infof("Skipping reconciling CSR '%s' because CSR request is not valid: %v", csr.ObjectMeta.Name, err)
+		return nil
+	}
+
+	klog.V(4).Infof("Approving CSR %s", csr.ObjectMeta.Name)
+	approvalCondition := certificatesv1beta1.CertificateSigningRequestCondition{
+		Type:   certificatesv1beta1.CertificateApproved,
+		Reason: "machine-controller NodeCSRApprover controller approved node serving cert",
+	}
+	csr.Status.Conditions = append(csr.Status.Conditions, approvalCondition)
+
+	if _, err := r.certClient.UpdateApproval(csr); err != nil {
+		return fmt.Errorf("failed to approve CSR %q: %v", csr.Name, err)
+	}
+
+	klog.Infof("Successfully approved CSR %s", csr.ObjectMeta.Name)
+	return nil
+}
+
+func isUsageInUsageList(usage certificatesv1beta1.KeyUsage, usageList []certificatesv1beta1.KeyUsage) bool {
+	for _, usageListItem := range usageList {
+		if usage == usageListItem {
+			return true
+		}
+	}
+	return false
+}
+
+// validateCSRRequest parses the CSR request and compares certificate request CommonName with CSR username, and
+// certificateRequest Organization with CSR groups.
+func validateCSRRequest(csr *certificatesv1beta1.CertificateSigningRequest) error {
+	csrBlock, rest := pem.Decode(csr.Spec.Request)
+	if csrBlock == nil {
+		return fmt.Errorf("no certificate request found for the given CSR")
+	}
+	if len(rest) != 0 {
+		return fmt.Errorf("found more than one PEM encoded block in the result")
+	}
+	csrRequest, err := x509.ParseCertificateRequest(csrBlock.Bytes)
+	if err != nil {
+		return err
+	}
+	// Validate Subject CommonName
+	if csrRequest.Subject.CommonName != csr.Spec.Username {
+		return fmt.Errorf("commonName '%s' is different then CSR username '%s'", csrRequest.Subject.CommonName, csr.Spec.Username)
+	}
+	// Validate Subject Organization
```

Comment:
As we discussed offline we can improve the validation to check SAN in following PRs